### ISSUE TITLE
Erroneous type error with Optional

### ISSUE THREAD

**Issue by 108anup** [microsoft/pyright#3470]
**Describe the bug**
Erroneous type error thrown when trying to assign a container with elements to a container with optional elements. 

**To Reproduce**
Steps to reproduce the behavior.
```
x: List[bool] = [False]
y: List[Optional[bool]] = x
```

**Observed behavior**
Get a type error:
```
Expression of type "list[bool]" cannot be assigned to declared type "List[bool | None]"
  TypeVar "_T@list" is invariant
    Type "bool" cannot be assigned to type "bool | None"
      Type cannot be assigned to type "None"
```

**Expected behavior**
Ideally there shouldn't be any type error here.

**Environment**
Python 3.9.7
Using pyright from pylance 2022.5.1 extension on vscode. (Not sure what pyright version it is using).

**Comment by erictraut** [comment]
This behavior is intended and is dictated by the Python type system documented in PEP 484. Since `list` is a mutable container type, its type parameter is invariant. That means you cannot assign a `list[bool]` to a `list[Optional[bool]]`. The two are not type compatible. For a simple explanation for why, consider the following extension to your code sample above:

```python
y.append(None)
```
This append operation violates the type of `x` because it refers to the same object that `y` refers to.

You can make this code type safe in a couple of different ways:
1. Declare `y` as a `Sequence[Optional[bool]]`. This is effectively an immutable form of list, which disallows `append` and other mutations.

```python
x: List[bool] = [False]
y: Sequence[Optional[bool]] = x
```

2. Allocate a new list so `x` and `y` do not point to the same underlying object.

```python
x: List[bool] = [False]
y: List[Optional[bool]] = list(x)
```

**Comment by erictraut** [comment]
For more information, you may find [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#generic-types) useful.
